### PR TITLE
Docs generation from `.sol` files.

### PR DESCRIPTION
**Pull request by dev-matan-tsuberi**
Automatic documentation from `.sol` files using [solcjs](https://github.com/ethereum/solc-js).
Use `yarn run docs:gen` to generate.

Changes: 
- Added `docs/gen/gen.js`
- `docs/ref/**.md` - generated files.

Future work:
- Intergrate with travis and automatically push to `daostack.wiki`.

### CODE REVIEW

File: docs/gen/gen.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+try{
+    main();
+    shell.echo('Done.');
+    shell.exit(0);
+}
+catch(e){
+    shell.echo(`An error occured: ${e.message}`);
+    shell.exit(1);
+}
```
Comment (orenyodfat) [inline docs/gen/gen.js]: no new line at the end of file

Comment (dev-matan-tsuberi) [reply docs/gen/gen.js]: `yarn lint` passes.

Comment (dev-matan-tsuberi) [reply docs/gen/gen.js]: but fixed.

File: docs/gen/gen.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    
+    // returns an `.md` string based on given data.
+    function render(file,contractName,abi,devdoc){
+        const events = abi.filter(x => x.type === 'event').sort((x,y) => x.name <= y.name);
+        const functions = abi.filter(x => x.type === 'function').sort((x,y) => x.name <= y.name);
+        const methods = devdoc.methods || {};
+        const title = devdoc.title || '';
+
+        /* This is very ugly, but in order for the generated markdown to be clean,
+           we cannot use any indentation which doesn't apprear in the `.md` file
```
Comment (orenyodfat) [inline docs/gen/gen.js]: > apprear 
spelling

Comment (dev-matan-tsuberi) [reply docs/gen/gen.js]: fixed.

File: docs/ref/VotingMachines/AbsoluteVote.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+3. **_avatar** *of type address* - an address to be sent as the payload to the _executable contract.
+4. **_executable** *of type address* - This contract will be executed when vote is over.
 TODO: Maybe we need to check the 0 < precReq <= 100 ??
+
+*Returns:*
+*Nothing*
+
+### *function* proposals
+**constant**
+**payable**
```
Comment (orenyodfat) [inline docs/ref/VotingMachines/AbsoluteVote.md]: payable ? It is ?

File: docs/gen/gen.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    }
+}
+
+try{
+    main();
+    shell.echo('Done.');
+    shell.exit(0);
+}
+catch(e){
+    shell.echo(`An error occured: ${e.message}`);
```
[the remark this answers is not in the document]
Comment (dev-matan-tsuberi) [reply docs/gen/gen.js]: fixed

File: docs/ref/VotingMachines/AbsoluteVote.md (hunk 3 above)
Comment (dev-matan-tsuberi) [reply docs/ref/VotingMachines/AbsoluteVote.md]: Thanks for spotting that! fixed.

**Comment by orenyodfat** [comment]
functions modifiers like public, private,internal,external ..are also nice to have.

**Comment by dev-matan-tsuberi** [comment]
This only documents the exposed interface. No private/internal functions, and the outside user does't care if it's external or public (except in regard to gas prices, which can be included regardless).

**Comment by dev-matan-tsuberi** [comment]
Re Inheritance: the docs do not indicate the parents, but they do include all inherited properties, functions etc...
Is this sufficient?

**Comment by orenyodfat** [comment]
> Code deposit upper limit: 861000 gas Executionas upper limit: 891 gas

???

**Comment by orenyodfat** [comment]
> Execution cost upper limit: Infinite gas constant | view

??

**Comment by orenyodfat** [comment]
> @orenyodfat Re Inheritance: the docs do not indicate the parents, but they do include all inherited properties, functions etc...
Is this sufficient?

I guess that it is nice to know from who the contract inherit (e.g is Ownble).

**Comment by dev-matan-tsuberi** [comment]
Re, gas estimates: those are real values generated by `solc`. the solidity compiler gives you upper bounds on the amount of gas a function costs. 
Code deposit is the cost caused b the size of the deployed code of the contract. Execution is the cost of executing the constructor/function. 
See: http://solidity.readthedocs.io/en/develop/using-the-compiler.html 
(look for `gasEstimates` in the output json)

**Comment by dev-matan-tsuberi** [comment]
Re, inheritance: can we drop this feature or maybe postpone it? I'm having trouble pulling this info from `solcjs`.

**Comment by orenyodfat** [comment]
> Re, gas estimates: those are real values generated by solc. the solidity compiler gives you upper bounds on the amount of gas a function costs.
Code deposit is the cost caused b the size of the deployed code of the contract. Execution is the cost of executing the constructor/function.
See: http://solidity.readthedocs.io/en/develop/using-the-compiler.html
(look for gasEstimates in the output json)

Yea, it will be nice to have that..lets try to format that more nice.

**Comment by orenyodfat** [comment]
> Re, inheritance: can we drop this feature or maybe postpone it? I'm having trouble pulling this info from solcjs.

Ok, lets issue that.

**Comment by orenyodfat** [comment]
> Execution cost upper limit: Infinite gas constant | view

What does Infinite gas constant  means ?

**Comment by dev-matan-tsuberi** [comment]
> What does Infinite gas constant means ?
No upper limit, the code could spend arbitrary amounts of gas. 
* Note: that doesn't mean that it will, but that the compiler could not provide a bound.

**Comment by dev-matan-tsuberi** [comment]
Added issues: 
https://github.com/daostack/daostack/issues/239 
https://github.com/daostack/daostack/issues/240

**Comment by orenyodfat** [comment]
> What does Infinite gas constant means ?
No upper limit, the code could spend arbitrary amounts of gas.
*Note: that doesn't mean that it will, but that the compiler could not provide a bound.

It is very confusing 
https://github.com/daostack/daostack/blob/docs-gen/docs/ref/controller/Avatar.md#function-sendether

How about replace Infinite gas with : unknown gas

**Comment by dev-matan-tsuberi** [comment]
How about now? https://github.com/daostack/daostack/blob/docs-gen/docs/ref/controller/Avatar.md#function-sendether

**Comment by dev-matan-tsuberi** [comment]
added table of contents https://github.com/daostack/daostack/blob/docs-gen/docs/ref/toc.md

**Comment by dev-matan-tsuberi** [comment]
Added feature: templates now include files from `docs/headers`, such that we can include manual documentation in the generated docs.

Check:
[`docs/headers/toc.md`](https://github.com/daostack/daostack/blob/docs-gen/docs/headers/toc.md)
[`docs/ref/toc.md`](https://github.com/daostack/daostack/blob/docs-gen/docs/ref/toc.md)
And:
[`docs/headers/globalConstraints/GlobalConstraintInterface.md`](https://github.com/daostack/daostack/blob/docs-gen/docs/headers/globalConstraints/GlobalConstraintInterface.md)
[`docs/ref/globalConstraints/GlobalConstraintInterface.md`](https://github.com/daostack/daostack/blob/docs-gen/docs/ref/globalConstraints/GlobalConstraintInterface.md)
For examples.

**Comment by orenyodfat** [comment]
Good work @dev-matan-tsuberi .
I guess that the headers (manuals) will be parts of future prs.right ?
So lets just add a place holders for each..something like ("comping soon ...") so any one looking on this repo will be aware of that..

**Comment by dev-matan-tsuberi** [comment]
@orenyodfat Thanks. I think it's best to not crowd the `headers` directory with too many files. We don't know necessarily that we would need a header for each and every contract (there are pretty trivial ones, and even then you still have the natspec). I gave `toc.md` and `GlobalConstraintInterface.md` as an example usage, Do you want to remove those in this PR (or maybe just leave `toc.md`)?

And yes, headers and other manual docs are coming soon in branch `docs` (view it [here](https://github.com/daostack/daostack/tree/docs/docs)).

**Comment by leviadam** [comment]
Good work @dev-matan-tsuberi.
I think the scirpts directory should not be called `gen`, but something more meaningful.
Further, it might be a good idea to put it into the `scripts` folder, and auto run it with a publish script.

**Comment by dev-matan-tsuberi** [comment]
@orenyodfat @leviadam 
All good now?
BTW: adding the publish script in `package.json` means we don't need to add it to travis right?

File: package.json [hunk 5]
```diff
@@ -20,7 +20,8 @@
     "solium": "solium --dir ./contracts/",
     "truffle": "truffle",
     "build": "rimraf build && truffle compile",
-    "docs:gen": "node docs/gen/gen.js"
+    "docs:gen": "node docs/scripts/gen.js",
+    "publish": "npm run docs:gen"
```
Comment (orenyodfat) [inline package.json]: publish should build truffle + get docs + publish npm .
Something like : 
"truffle build && npm run docs:gen && network --clean && npm publish"

Comment (dev-matan-tsuberi) [reply package.json]: sorry, thought it was automatic.

File: package.json [hunk 6]
```diff
@@ -21,7 +21,7 @@
     "truffle": "truffle",
     "build": "rimraf build && truffle compile",
     "docs:gen": "node docs/scripts/gen.js",
-    "publish": "npm run docs:gen"
+    "publish": "truffle build && npm run docs:gen && network --clean && npm publish"
```
Comment (orenyodfat) [inline package.json]: > "truffle build && npm run docs:gen && network --clean && npm publish"

Did you tested this syntax ?
It should be more like ...
> "truffle compile && npm run docs:gen && truffle network --clean && npm publish"

Comment (dev-matan-tsuberi) [reply package.json]: Well, I'm not really familiar with how we should deploy Arc. 
As it's not really relevant to this PR, I think we should add a publish script on a different PR and then add `docs:gen`. 
Agreed?

Comment (orenyodfat) [reply package.json]: ok. please remove the publish.
I will add that later.

**Comment by orenyodfat** [comment]
This PR was totally reverted from master.
Needed to squashed all commits.

**Pull request merged**